erts/epmd/src/epmd_srv.c

```diff
@@ -340,13 +360,39 @@ void run(EpmdVars *g)
 #endif /* HAVE_SYSTEMD_DAEMON */
   for (i = 0; i < num_sockets; i++)
     {
-      if ((listensock[i] = socket(FAMILY,SOCK_STREAM,0)) < 0)
+      struct sockaddr *sa = (struct sockaddr *)&iserv_addr[i];
+#if defined(EPMD6)
+      size_t salen = (sa->sa_family == AF_INET6 ?
+              sizeof(struct sockaddr_in6) :
+              sizeof(struct sockaddr_in));
+#else
+      size_t salen = sizeof(struct sockaddr_in);
+#endif
+
+      if ((listensock[i] = socket(sa->sa_family,SOCK_STREAM,0)) < 0)
 	{
-	  dbg_perror(g,"error opening stream socket");
-	  epmd_cleanup_exit(g,1);
+	  switch (errno) {
+	      case EAFNOSUPPORT:
+	      case EPROTONOSUPPORT:
+	          continue;
```

Comment:
In this error case, a negative file descriptor is now saved at listensock[i], which leads to a buffer overflow later on at https://github.com/msantos/otp/blob/epmd-IPv6-node-reg2/erts/epmd/src/epmd_srv.c#L498 when calling `FD_ISSET(listensock[i],&read_mask)`.

To reproduce this:

1. Completely disable IPv6 with ipv6.disable=1 on the kernel command line
2. Build with fortified code (RHEL/CentOS default): CFLAGS='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic -fno-strict-aliasing'

And you get:

``` plain
[jeckersb@centos7 ~]$ ./epmd -d
epmd: Fri Feb 26 15:28:53 2016: epmd running - daemon = 0
*** buffer overflow detected ***: ./epmd terminated
======= Backtrace: =========
/lib64/libc.so.6(__fortify_fail+0x37)[0x7fd54f093b37]
/lib64/libc.so.6(+0x10bcf0)[0x7fd54f091cf0]
/lib64/libc.so.6(+0x10daa7)[0x7fd54f093aa7]
./epmd[0x40387a]
./epmd[0x401765]
/lib64/libc.so.6(__libc_start_main+0xf5)[0x7fd54efa7b15]
./epmd[0x401b41]
======= Memory map: ========
00400000-0040c000 r-xp 00000000 fd:03 407457                             /home/jeckersb/epmd
0060b000-0060c000 r--p 0000b000 fd:03 407457                             /home/jeckersb/epmd
0060c000-0060d000 rw-p 0000c000 fd:03 407457                             /home/jeckersb/epmd
01c8d000-01cae000 rw-p 00000000 00:00 0                                  [heap]
7fd54ed70000-7fd54ed85000 r-xp 00000000 fd:03 136225                     /usr/lib64/libgcc_s-4.8.5-20150702.so.1
7fd54ed85000-7fd54ef84000 ---p 00015000 fd:03 136225                     /usr/lib64/libgcc_s-4.8.5-20150702.so.1
7fd54ef84000-7fd54ef85000 r--p 00014000 fd:03 136225                     /usr/lib64/libgcc_s-4.8.5-20150702.so.1
7fd54ef85000-7fd54ef86000 rw-p 00015000 fd:03 136225                     /usr/lib64/libgcc_s-4.8.5-20150702.so.1
7fd54ef86000-7fd54f13c000 r-xp 00000000 fd:03 135341                     /usr/lib64/libc-2.17.so
7fd54f13c000-7fd54f33c000 ---p 001b6000 fd:03 135341                     /usr/lib64/libc-2.17.so
7fd54f33c000-7fd54f340000 r--p 001b6000 fd:03 135341                     /usr/lib64/libc-2.17.so
7fd54f340000-7fd54f342000 rw-p 001ba000 fd:03 135341                     /usr/lib64/libc-2.17.so
7fd54f342000-7fd54f347000 rw-p 00000000 00:00 0 
7fd54f347000-7fd54f448000 r-xp 00000000 fd:03 135349                     /usr/lib64/libm-2.17.so
7fd54f448000-7fd54f647000 ---p 00101000 fd:03 135349                     /usr/lib64/libm-2.17.so
7fd54f647000-7fd54f648000 r--p 00100000 fd:03 135349                     /usr/lib64/libm-2.17.so
7fd54f648000-7fd54f649000 rw-p 00101000 fd:03 135349                     /usr/lib64/libm-2.17.so
7fd54f649000-7fd54f64c000 r-xp 00000000 fd:03 135347                     /usr/lib64/libdl-2.17.so
7fd54f64c000-7fd54f84b000 ---p 00003000 fd:03 135347                     /usr/lib64/libdl-2.17.so
7fd54f84b000-7fd54f84c000 r--p 00002000 fd:03 135347                     /usr/lib64/libdl-2.17.so
7fd54f84c000-7fd54f84d000 rw-p 00003000 fd:03 135347                     /usr/lib64/libdl-2.17.so
7fd54f84d000-7fd54f84f000 r-xp 00000000 fd:03 135375                     /usr/lib64/libutil-2.17.so
7fd54f84f000-7fd54fa4e000 ---p 00002000 fd:03 135375                     /usr/lib64/libutil-2.17.so
7fd54fa4e000-7fd54fa4f000 r--p 00001000 fd:03 135375                     /usr/lib64/libutil-2.17.so
7fd54fa4f000-7fd54fa50000 rw-p 00002000 fd:03 135375                     /usr/lib64/libutil-2.17.so
7fd54fa50000-7fd54fa71000 r-xp 00000000 fd:03 135334                     /usr/lib64/ld-2.17.so
7fd54fc60000-7fd54fc64000 rw-p 00000000 00:00 0 
7fd54fc6f000-7fd54fc71000 rw-p 00000000 00:00 0 
7fd54fc71000-7fd54fc72000 r--p 00021000 fd:03 135334                     /usr/lib64/ld-2.17.so
7fd54fc72000-7fd54fc73000 rw-p 00022000 fd:03 135334                     /usr/lib64/ld-2.17.so
7fd54fc73000-7fd54fc74000 rw-p 00000000 00:00 0 
7fff581e1000-7fff58202000 rw-p 00000000 00:00 0                          [stack]
7fff58306000-7fff58308000 r-xp 00000000 00:00 0                          [vdso]
ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]
Aborted
```